Submit STAC Items via STAC Ingestor

alukach opened the pull request:
## What I'm changing

This PR updates the data pipelines to no longer directly insert STAC items into pgSTAC and instead go through the VEDA STAC Ingestor.  This brings in the following benefits:

1. Removes risks associated with overrunning database with too many connections in the event that we are running many concurrent data pipelines.
2. Removes need to deploy AWS Lambda into the same VPC as the RDS database running pgSTAC.  This means that the data pipelines can run in a region or AWS account that is different from the pgSTAC region/account.

## How I did it

The general idea is to drop the `db-write` task and replace it with a `submit-stac` task.  Rather than connecting to pgSTAC, we'll instead send POST requests for each individual STAC Item to the STAC Ingestor.

### Relevant Commits

- Rename `db-write` task to `submit-stac` (f12932c1244a2bbf50885ed1932a7f26b633fb5c)
- Remove concept of VPC ID and Security Group, as they are no longer needed when we don't need direct DB access (2ef0ec62c7de2972d127bb06c2c48a1d61ed2e99)
- Rework handler to make authenticate POST requests to the STAC Ingestor API (17b0bfb318f67ea7f8acb5acae18948a4a981f1f)
- Remove `build-ndjson` task, instead iterate over the array of STAC items and individually call `submit-stac` (28ed3dd3dde3198a27c8fbb0b5f06ae1922844d8)

### Along the way...

Along the way, I noticed a few opportunities for cleanup.  These may be contentious, you're welcome to challenge any of these changes.

- Move `deploy/app.py:integrate_to_lambdas()` to `deploy/cdk/lambda_stack.py:LambdaStack.grant_execution_privileges()` (3bbce17a70e81180a1a3ff861a2a3a4f301a54d1)
- Remove `IamPolicies` class in favor of leaning on CDK's S3 bucket permissions helpers (10b3de6de2c1e8c86e5df28877188e6cbd9985a5). This allows us to lessen the amount of custom code in the project.
- Use attributes (e.g. `self.trigger_cogify_lambda`) rather than a dict attribute (e.g. `self.lambdas = {'trigger_cogify_lambda': trigger_cogify_lambda}`) (eab54fd6a32f7a126ecd4c28ec814d1e33bb43f3, 725f83361ceaaba0baa1be7bd03fe2846f987cd4). This allows MyPy & text editors to easily follow the class attributes and to provide typing assistance (e.g. VSCode knows that `lambda_stack.trigger_cogify_lambda` is of type `cdk_lambdas.Function` and will provide code completion via IntelliSense)
- Reorg our step functions into a separate method for each workflow (5fbda1f4286955d43f074341d4dfb55dc53d955c). The goal here is to make it simpler to understand each workflow's tasks and flow.

## How you can test it

You need two new environment variables to deploy this:

* `COGNITO_APP_SECRET`, this is a secret created by [veda-auth](https://github.com/NASA-IMPACT/veda-auth).  You can use `veda-auth-stack-alukach/veda-workflows` when deploying to the VEDA Dev AWS Account (`853558080719`).  To manually review the secret:
    ```
    aws secretsmanager get-secret-value --secret-id veda-auth-stack-alukach/veda-workflows  --output text --query SecretString
    ```
* `STAC_INGESTOR_URL`, this is the API of our [veda-stac-ingestor](https://github.com/NASA-IMPACT/veda-stac-ingestor/).  You can use `https://52hxaagqch.execute-api.us-west-2.amazonaws.com/alukach`.
---

Closes #163
Closes #165

slesaad commented:
Hi @alukach.
First of all, I really appreciate you making the "along the way..." changes. Those are awesome and make the codebase so much better. 🙇 

I do notice tho that you've completely removed the `build-ndjson` lambda, which actually would build the "stac-item" that's supposed to be sent to the `submit-stac` lambda. It used `rio_stac.stac.create_stac_item()` function to create an stac-item from a `.tif` file, the filename, and user input like `datetime_range`.

The output from `data-transfer` task looks something like:
```json
[{
    "filename_regex": "^(.*)2022(.*).tif$",
    "datetime_range": null,
    "properties": null,
    "collection": "lis-tws-anomaly",
    "s3_filename": "s3://veda-data-store-staging/EIS/COG/LIS_TWS_ANOMALY/Anomaly_TWS_20120224.cog.tif",
    "href": "s3://veda-data-store-staging/EIS/COG/LIS_TWS_ANOMALY/Anomaly_TWS_20120224.cog.tif",
    "id": "EIS/COG/LIS_TWS_ANOMALY/Anomaly_TWS_20120224.cog.tif",
    "upload": false
  },
...
]
```

while the expected input for `submit-stac` looks like:
```json
{
    "stac_file_url": "example.ndjson",
    # or
    "stac_item": {},
    "type": "collections",
}
```

so looks like we're missing an intermediate step of generating the stac item from the tif file.

alukach commented on lambdas/build-stac/src/handler.py:
```diff
@@ -0,0 +1,163 @@
+    stac_item = parse_obj_as(Event, event).as_stac()
+    stac_dict = stac_item.to_dict()
+
+    # # TODO: Remove this hack
+    # if s3_url := event.get("data_url"):
+    #     stac_dict["assets"]["cog_default"]["href"] = s3_url
```
Not sure if this should stay or go.

alukach commented:
Thanks @slesaad, good catch!  

I've taken the `build-ndjson` step and converted it into the `build-stac` step.  It's slightly different though.  

### Before

The `build-ndjson` step would take many messages and builds a single NDJSON.

### After

We are using the [Map task](https://docs.aws.amazon.com/step-functions/latest/dg/amazon-states-language-map-state.html) to iterate over the incoming array of messages. This runs each message through the `build-stac` step and then the `submit-stac` step.  The reason that I opted to build each STAC item individually was due to the fact that the STAC Ingestor does _not_ support bulk submissions, so I recommend we only submit a single STAC Item per lambda so that if any item fails, it does not prevent other items from being submitted.  When the STAC item is built, it is written to S3 and that s3 key is passed to the `submit-stac` step.  This means we're doing 1 S3 PUT operation for every STAC item, which has the downside of causing many writes which could increase costs.  The upside of writing the STAC item to S3 is that we don't need to worry about STAC items being too large to fit into AWS Step Functions messages ([256KB](https://aws.amazon.com/about-aws/whats-new/2020/09/aws-step-functions-increases-payload-size-to-256kb/#:~:text=AWS%20Step%20Functions%20now%20supports,the%20limit%20was%2032%2C768%20characters.)). I'm not sure if this is the right call, if we are confident that STAC items will never exceed 256KB then the system would be much more cost effective to pass the STAC item directly from the output of the `build-stac` task and into the input of `submit-stac` task.  Curious to hear what you think @slesaad & @xhagrg.

### Along the way

#### Refactored filename parsing

I refactored the code that extracts datetimes from the filename.  I found it really challenging to understand how the single `regex` and multiple `format` values worked together.  Instead, I broke it into separate regex & format pairs:

<details>

```py

from datetime import datetime

def old_extract(filename, datetime_range):
    """
    Extracts start, end, or single date string from filename
    and convert it to iso format datetime.
    """
    DATE_REGEX_STRATEGIES = [
        {
            "regex": re.compile("(_\d{4}-\d{2}-\d{2})|(_\d{8})|(_\d{6})|(_\d{4})"),
            "format": ["%Y-%m-%d", "%Y%m%d", "%Y%m", "%Y"],
        }
    ]

    dates = []

    for regex_dict in DATE_REGEX_STRATEGIES:
        internal_dates = regex_dict["regex"].findall(filename)
        internal_dates = [
            (index, date)
            for internal_date in internal_dates
            for index, date in enumerate(internal_date)
            if date
        ]
        for index, internal_date in internal_dates:
            if internal_date := internal_date.replace("_", ""):
                dates += [
                    datetime.strptime(internal_date, regex_dict["format"][index])
                ]
    dates.sort()

    return dates


def new_extract(filename, datetime_range):
    DATE_REGEX_STRATEGIES = [
        (r'_(\d{4}-\d{2}-\d{2})', "%Y-%m-%d"),
        (r'_(\d{8})', "%Y%m%d"),
        (r'_(\d{6})', "%Y%m"),
        (r'_(\d{4})', "%Y"),
    ]

    dates = []
    for (pattern, dateformat) in DATE_REGEX_STRATEGIES:
        dates_found = re.compile(pattern).findall(filename)
        if not dates_found:
            continue

        for date_str in dates_found:
            dates.append(datetime.strptime(date_str, dateformat))
    dates.sort()
    return dates

assert old_extract('s3://climatedashboard-data/OMSO2PCA/OMSO2PCA_LUT_SCD_2005_2006_2007.tif', None) == [datetime(2005, 1, 1, 0, 0),
 datetime(2006, 1, 1, 0, 0),
 datetime(2007, 1, 1, 0, 0)]
assert new_extract('s3://climatedashboard-data/OMSO2PCA/OMSO2PCA_LUT_SCD_2005_2006_2007.tif', None) == [datetime(2005, 1, 1, 0, 0),
 datetime(2006, 1, 1, 0, 0),
 datetime(2007, 1, 1, 0, 0)]

```

</details>

#### Removed `mode == "stac"` code

I don't believe this `mode == "stac"` code was ever used (it would fail at `requests.get()` if it were):

https://github.com/NASA-IMPACT/veda-data-pipelines/blob/27eb257b13ba4032bdcb96d8eb2e7c65a34092e5/lambdas/build-ndjson/handler.py#L259-L266

#### Remove any creation of `assets`

We have some logic to generate an `assets` dict and to pass it to the `create_item()` function:

https://github.com/NASA-IMPACT/veda-data-pipelines/blob/27eb257b13ba4032bdcb96d8eb2e7c65a34092e5/lambdas/build-ndjson/handler.py#L161-L177

However, in `create_item()`, we never make use of the `assets` argument.  For this reason, I removed any logic around generating a dictionary of assets.

https://github.com/NASA-IMPACT/veda-data-pipelines/blob/27eb257b13ba4032bdcb96d8eb2e7c65a34092e5/lambdas/build-ndjson/handler.py#L69-L101

#### Tests!

I added some tests to validate that the regex works as expects.  You should be able to run `pytest lambdas` from the root of the project.  Eventually, we should integrate this into Github CI.

abarciauskas-bgse commented:
This is great @alukach thank you!
If I understand correct, there is no more use of ndjson right? In which case I think we can remove lambdas/submit-stac/hlss30_stac_example.ndjson and rename the `ndjson_bucket` variable in https://github.com/NASA-IMPACT/veda-data-pipelines/blob/feature/submit-stac-items-via-ingestor/deploy/cdk/lambda_stack.py#L68? Minor things

alukach commented:
@abarciauskas-bgse Yeah, no more ndjson in this setup so we could remove `lambdas/submit-stac/hlss30_stac_example.ndjson`, however we still need a bucket to store the STAC items.  Being that this codebase does not create the bucket, but instead imports the bucket, I didn't want to change its name (forcing us to create a new bucket during the next deployment).  As I mentioned above, I am unsure if we should even be writing the stac messages to S3:

> I'm not sure if this is the right call, if we are confident that STAC items will never exceed 256KB then the system would be much more cost effective to pass the STAC item directly from the output of the build-stac task and into the input of submit-stac task.

slesaad commented:
@alukach i think it's safe to assume that one stac item will not cross the 256KB threshold (i checked the size of a typical item, it was 2KB). so we could probably pass the item directly in the output rather than writing to s3.

other than that, everything looks great to me!

xhagrg commented on deploy/cdk/lambda_stack.py:
```diff
@@ -157,58 +142,40 @@ def _python_lambda(self, name, directory, env=None, timeout_seconds=900, **kwarg
-    def lambda_sg(self):
-        return self._lambda_sg
+        for bucket in [internal_bucket, *external_buckets]:
+            bucket.grant_read(self.s3_discovery_lambda.role)
+            bucket.grant_read(self.build_stac_lambda.role)
+            if self.data_transfer_lambda:
+                bucket.grant_read(self.data_transfer_lambda.role)
```
Is this not updating the bucket policy? It should be fine for the non-NASA deployments, but we won't be able to use it as is in a NASA system. Updating bucket policy is not available to us.

alukach replied:
I agree that it does look like this is an action that would affect the `bucket` object, [the docs](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_s3.Bucket.html#grantwbrreadidentity-objectskeypattern) validate that it alters the provided principal:

> Grant read permissions for this bucket and it's contents to an IAM principal (Role/Group/User).

xhagrg commented:
It might also be worth a to remove the queues before generating stacs.

alukach commented on lambdas/build-stac/handler.py:
```diff
@@ -0,0 +1,74 @@
+    parsed_event = parse_obj_as(events.SupportedEvent, event)
+    stac_item = stac.generate_stac(parsed_event).to_dict()
+
+    # TODO: Remove this hack
+    if s3_url := event.get("data_url"):
+        stac_item["assets"]["cog_default"]["href"] = s3_url
```
Okay, I think this PR is good to go except for this one final question.  What is the story behind this?  Should I remove it?

@abarciauskas-bgse @slesaad @xhagrg

slesaad replied:
yes!

alukach replied:
Ok, removed!

alukach commented on lambdas/build-stac/handler.py:
```diff
@@ -0,0 +1,74 @@
+
+    parsed_event = parse_obj_as(events.SupportedEvent, event)
+    stac_item = stac.generate_stac(parsed_event).to_dict()
+
+    # TODO: Remove this hack
+    if s3_url := event.get("data_url"):
+        stac_item["assets"]["cog_default"]["href"] = s3_url
```
```suggestion
```
